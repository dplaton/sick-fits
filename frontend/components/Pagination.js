import React, {Component} from 'react'
import gql from 'graphql-tag'
import {Query} from 'react-apollo';
import Head from 'next/head';
import Link from 'next/link'

import PaginationStyles from './styles/PaginationStyles'
import {perPage} from '../config'

export const PAGINATION_QUERY = gql `
    query PAGINATION_QUERY {
        itemsConnection {
            aggregate {
                count
            }
        }
    }
`

const Pagination = (props) => (
    <Query query={PAGINATION_QUERY}>
        {({data, loading, error}) => {
            if (loading) return <p>Loading...</p>;
            const itemsCount = data.itemsConnection.aggregate.count
            const numPages = Math.ceil(itemsCount / perPage);
            const {page} = props;
            return (
                <PaginationStyles data-test="pagination">
                    <Head>
                        <title>Sick Fits | page {page}
                            of {numPages}</title>
                    </Head>
                    <Link
                        href={{
                        pathname: 'items',
                        query: {
                            page: page - 1
                        }
                    }}>
                        <a className='prev' aria-disabled={page <= 1}>Prev</a>
                    </Link>
                    <p>{page}&nbsp; of <span className="total-pages">{numPages}</span></p>
                    <p>{itemsCount}&nbsp; items total</p>
                    <Link
                        href={{
                        pathname: 'items',
                        query: {
                            page: page + 1
                        }
                    }}>
                        <a className='next' aria-disabled={page >= numPages}>Next</a>
                    </Link>
                </PaginationStyles>
            )
        }}
    </Query>
)

export default Pagination
