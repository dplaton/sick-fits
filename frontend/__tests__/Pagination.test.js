import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import Router from "next/router";

import Pagination, { PAGINATION_QUERY } from "../components/Pagination";

// mock the router
Router.router = {
    push() { },
    prefetch() { }
}

const mockZeroElements = [
    {
        request: { query: PAGINATION_QUERY },
        result: { data: { itemsConnection: { aggregate: { count: 0 } } } }
    }
];

function makeMocksFor(count) {
    return [
        {
            request: {
                query: PAGINATION_QUERY
            },
            result: {
                data: {
                    itemsConnection: {
                        __typename: "aggregate",
                        aggregate: { 
                            __typename: "count", 
                            count 
                        }
                    }
                }
            }
        }
    ];
}

describe("<Pagination />", () => {
    if ("displays a loading message", async ()=> {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(1)}>
                <Pagination page={1}/>
            </MockedProvider>
        );
        expect(wrapper.text()).toContain('Loading...');
    });

    it("shows a single item when there are 18 elements", async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        expect(wrapper.find('.total-pages').text()).toEqual('5');
        const pagination = wrapper.find('div[data-test="pagination"]');
        expect(toJSON(pagination)).toMatchSnapshot();
    });

    it("disables the 'Next' button at last page", async() => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={5}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        const prevLink = wrapper.find('a.prev');
        const nextLink = wrapper.find('a.next')
        expect(nextLink.prop("aria-disabled")).toBe(true);
        expect(prevLink.prop("aria-disabled")).toBe(false);
    });

    it("disables the 'Prev' button at first page", async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={1}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        const prevLink = wrapper.find('a.prev');
        const nextLink = wrapper.find('a.next')
        expect(prevLink.prop("aria-disabled")).toBe(true);
        expect(nextLink.prop("aria-disabled")).toBe(false);
    });

    it("enables both 'Prev' and 'Next' buttons at a middle page", async () => {
        const wrapper = mount(
            <MockedProvider mocks={makeMocksFor(18)}>
                <Pagination page={3}/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();
        const prevLink = wrapper.find('a.prev');
        const nextLink = wrapper.find('a.next')
        expect(prevLink.prop("aria-disabled")).toBe(false);
        expect(nextLink.prop("aria-disabled")).toBe(false);
    });
});
