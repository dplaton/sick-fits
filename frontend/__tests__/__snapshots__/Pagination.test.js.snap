// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> shows a single item when there are 18 elements 1`] = `
<div
  className="PaginationStyles-aduuar-0 ewJsCc"
  data-test="pagination"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 0,
        },
      }
    }
  >
    <a
      aria-disabled={true}
      className="prev"
      href="items?page=0"
      onClick={[Function]}
    >
      Prev
    </a>
  </Link>
  <p>
    1
      of 
    <span
      className="total-pages"
    >
      5
    </span>
  </p>
  <p>
    18
      items total
  </p>
  <Link
    href={
      Object {
        "pathname": "items",
        "query": Object {
          "page": 2,
        },
      }
    }
  >
    <a
      aria-disabled={false}
      className="next"
      href="items?page=2"
      onClick={[Function]}
    >
      Next
    </a>
  </Link>
</div>
`;
